File: PlayerAssistant.Tests/ProtectedDataNegativeSpaceTests.cs
using PlayerAssistant;

namespace PlayerAssistant.Tests;

internal static class ProtectedDataNegativeSpaceTests
{
    internal static void ObservabilityArtifactsNeverContainProtectedData()
    {
        var canaries = new[]
        {
            "fixture-password-7f2d", "fixture-bearer-8a31", "fixture-admin-key-9c42",
            "fixture-cookie-ad3e", "fixture-storage-state-be54", "fixture-private-path-cf65",
            "fixture-response-body-dg76"
        };
        var inventory = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["logging"] = "level=error message=login failed password=fixture-password-7f2d",
            ["errors"] = "exception=UnauthorizedException token=fixture-bearer-8a31",
            ["metrics"] = "metric=auth_failure admin_key=fixture-admin-key-9c42",
            ["diagnostics"] = "Cookie: session=fixture-cookie-ad3e storage_state=fixture-storage-state-be54",
            ["crash_reports"] = "path=C:\\Users\\Bryan\\private\\fixture-private-path-cf65 response_body=fixture-response-body-dg76",
            ["http_responses"] = "Authorization: " + "Bearer " + "fixture-bearer-8a31 body=fixture-response-body-dg76",
            ["browser_console"] = "error: cookie=fixture-cookie-ad3e storage state=fixture-storage-state-be54",
            ["ci_artifacts"] = "private_path=C:\\Users\\Bryan\\secret\\fixture-private-path-cf65",
            ["generated_bundles"] = "{\"password\":\"fixture-password-7f2d\",\"admin_key\":\"fixture-admin-key-9c42\"}"
        };
        if (inventory.Count != 9) throw new InvalidOperationException("protected-data inventory is incomplete");
        foreach (var (kind, raw) in inventory)
        {
            var redacted = SensitiveTextRedactionUtility.Redact(raw);
            foreach (var canary in canaries)
                if (redacted.Contains(canary, StringComparison.Ordinal))
                    throw new InvalidOperationException($"{kind} leaked protected fixture data");
            if (!redacted.Contains(SensitiveTextRedactionUtility.RedactedValue, StringComparison.Ordinal))
                throw new InvalidOperationException($"{kind} did not emit a redaction marker");
        }
    }

    internal static void SafeIdentifiersRemainInObservabilityArtifacts()
    {
        const string safe = "correlation_id=fixture-correlation-001 endpoint=/v1/health status=401";
        var redacted = SensitiveTextRedactionUtility.Redact(safe);
        if (!string.Equals(safe, redacted, StringComparison.Ordinal))
            throw new InvalidOperationException("safe observability identifiers were altered");
    }
}

File: web-deploy/tests/protected-data-negative-space-tests.php
<?php
declare(strict_types=1);

require_once __DIR__ . '/../player-assistant-broker/CorrelationContext.php';

$canaries = [
    'fixture-password-7f2d', 'fixture-bearer-8a31', 'fixture-admin-key-9c42',
    'fixture-cookie-ad3e', 'fixture-storage-state-be54', 'fixture-private-path-cf65',
    'fixture-response-body-dg76',
];
$artifacts = [
    'logging' => 'password=fixture-password-7f2d',
    'errors' => 'Authorization: Bearer fixture-bearer-8a31',
    'metrics' => 'admin_key=fixture-admin-key-9c42',
    'diagnostics' => 'Cookie: session=fixture-cookie-ad3e storage_state=fixture-storage-state-be54',
    'crash_reports' => 'path=C:\\Users\\Bryan\\fixture-private-path-cf65',
    'http_responses' => 'body=fixture-response-body-dg76',
    'browser_console' => 'token=fixture-bearer-8a31',
    'ci_artifacts' => 'private_path=C:\\Users\\Bryan\\fixture-private-path-cf65',
    'generated_bundles' => 'password=fixture-password-7f2d',
];
foreach ($artifacts as $kind => $raw) {
    $redacted = CorrelationContext::redact($raw);
    foreach ($canaries as $canary) {
        if (str_contains($redacted, $canary)) throw new RuntimeException("$kind leaked protected fixture data");
    }
    if (!str_contains($redacted, '[REDACTED]')) throw new RuntimeException("$kind lacked redaction marker");
}
$safe = 'correlation_id=fixture-correlation-001 endpoint=/v1/health status=401';
if (CorrelationContext::redact($safe) !== $safe) throw new RuntimeException('safe identifiers changed');
echo "Protected-data negative-space PHP tests passed.\n";
